keras_cv/models/task.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2023 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Base class for Task models."""
+
+import os
+
+from tensorflow import keras
+
+from keras_cv.utils.python_utils import classproperty
+from keras_cv.utils.python_utils import format_docstring
+
+
+@keras.utils.register_keras_serializable(package="keras_nlp")
+class Task(keras.Model):
+    """Base class for Task models."""
+
+    @property
+    def backbone(self):
+        """A `keras.Model` instance providing the backbone submodel."""
+        return self._backbone
+
+    @backbone.setter
+    def backbone(self, value):
+        self._backbone = value
+
+    def get_config(self):
+        # Don't chain to super here. The default `get_config()` for functional
+        # models is nested and cannot be passed to our Task constructors.
+        return {
+            "backbone": keras.layers.serialize(self.backbone),
+            "name": self.name,
+            "trainable": self.trainable,
+        }
+
+    @classmethod
+    def from_config(cls, config):
+        # The default `from_config()` for functional models will return a
+        # vanilla `keras.Model`. We override it to get a subclass instance back.
+        if "backbone" in config and isinstance(config["backbone"], dict):
+            config["backbone"] = keras.layers.deserialize(config["backbone"])
+        return cls(**config)
+
+    @classproperty
+    def presets(cls):
+        """Dictionary of preset names and configurations."""
+        return {}
+
+    @classproperty
+    def presets_with_weights(cls):
+        """Dictionary of preset names and configurations that include weights."""
+        return {}
+
+    @classproperty
+    def backbone_presets(cls):
+        """Dictionary of preset names and configurations for compatible backbones."""
+        return {}
+
+    @classmethod
+    def from_preset(
+        cls,
+        preset,
+        load_weights=True,
```

Comment:
Does a preset definitionally include weights?
Or can a preset without any pre-trained weights exist?

Reply:
For KerasNLP no, but for our backbones @fchollet wanted no-weights backbones (`"resnet50_v2"`). Even if the Task has weights, in some cases you just want the config so we added the `load_weights` flag.

Let me change the default to `None` to match my backbone PR so that we don't get errors by default for these no-weights presets.

Reply:
For KerasNLP no, but for our backbones @fchollet wanted no-weights backbones (`"resnet50_v2"`). Even if the Task has weights, in some cases you just want the config so we added the `load_weights` flag.

Let me change the default to `None` to match my backbone PR so that we don't get errors by default for these no-weights presets.